### PR TITLE
Implement st_simplify

### PR DESCRIPTION
**Pull request by tdikland**
Implementation of the `st_simplify` spatial function.

Concrete `MosaicGeometry.simplify(tol)` was already implemented, however the ESRI variant does ignore the specified tolerance. This pull request does not change that, hence the outcome of this function depend on the geometry api bound to the context.

### CODE REVIEW

File: python/test/test_functions.py [hunk 1]
```diff
@@ -195,8 +196,3 @@ def test_aggregation_functions(self):
                 "comparison_intersection"
             ]
         )
-
-    def test_union_agg(self):
```
Comment (edurdevic) [inline python/test/test_functions.py:199]: Why did you remove this test?

File: src/test/scala/com/databricks/labs/mosaic/expressions/geometry/ST_SimplifyTest.scala [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+    private val codegenOnly =
+        withSQLConf(
+          SQLConf.ADAPTIVE_EXECUTION_ENABLED.key -> "false",
+          SQLConf.WHOLESTAGE_CODEGEN_ENABLED.key -> "true",
+          SQLConf.CODEGEN_FACTORY_MODE.key -> CodegenObjectFactoryMode.CODEGEN_ONLY.toString
+        ) _
+
+    test("Testing stSimplify (H3, JTS) NO_CODEGEN") { noCodegen { simplifyBehavior(H3IndexSystem, JTS) } }
+    test("Testing stSimplify (H3, ESRI) NO_CODEGEN") { noCodegen { simplifyBehavior(H3IndexSystem, ESRI) } }
```
Comment (edurdevic) [inline src/test/scala/com/databricks/labs/mosaic/expressions/geometry/ST_SimplifyTest.scala:26]: Here also we don't need to test across grid systems

File: docs/source/api/spatial-functions.rst [hunk 3]
```diff
@@ -1209,6 +1209,61 @@ st_translate
     |MULTIPOINT ((20 35), (50 25), (30 15), (40 5))|
     +----------------------------------------------+
 
+st_simplify
+***********
+
+.. function:: st_simplify(geom, tol)
+
+    Returns the simplified geometry.
```
Comment (edurdevic) [inline docs/source/api/spatial-functions.rst:1217]: I would add a note that the tolerance is ignored with ESRI geometry engine.

**Review by edurdevic — CHANGES_REQUESTED** [review]
It looks good!
I only added a few NIT comments, and the pipeline is failing the python tests :)
Once those are resolved we can merge

File: src/test/scala/com/databricks/labs/mosaic/expressions/geometry/ST_SimplifyTest.scala (hunk 2 above)
Comment (tdikland) [reply src/test/scala/com/databricks/labs/mosaic/expressions/geometry/ST_SimplifyTest.scala:26]: This only tests against the (real) H3 grid index system. I'm working on a separate PR to use a mock index system for these kind of tests.

File: src/main/scala/com/databricks/labs/mosaic/functions/MosaicContext.scala [hunk 4]
```diff
@@ -506,6 +511,8 @@ class MosaicContext(indexSystem: IndexSystem, geometryAPI: GeometryAPI) extends
         def st_scale(geom1: Column, xd: Column, yd: Column): Column =
             ColumnAdapter(ST_Scale(geom1.expr, xd.expr, yd.expr, geometryAPI.name))
         def st_setsrid(geom: Column, srid: Column): Column = ColumnAdapter(ST_SetSRID(geom.expr, srid.expr, geometryAPI.name))
+        def st_simplify(geom: Column, tolerance: Double): Column =
```
[the remark this answers is not in the document]
Comment (tdikland) [reply src/main/scala/com/databricks/labs/mosaic/functions/MosaicContext.scala:516]: Agree!

File: src/main/scala/com/databricks/labs/mosaic/functions/MosaicContext.scala [hunk 5]
```diff
@@ -511,6 +511,8 @@ class MosaicContext(indexSystem: IndexSystem, geometryAPI: GeometryAPI) extends
         def st_scale(geom1: Column, xd: Column, yd: Column): Column =
             ColumnAdapter(ST_Scale(geom1.expr, xd.expr, yd.expr, geometryAPI.name))
         def st_setsrid(geom: Column, srid: Column): Column = ColumnAdapter(ST_SetSRID(geom.expr, srid.expr, geometryAPI.name))
+        def st_simplify(geom: Column, tolerance: Column): Column =
```
Comment (edurdevic) [inline src/main/scala/com/databricks/labs/mosaic/functions/MosaicContext.scala:514]: This is now not covered by unit tests.
Could you please cover this one as well?

**Review by edurdevic — APPROVED** [review]
Looks good! Just a small test coverage regression

File: src/main/scala/com/databricks/labs/mosaic/functions/MosaicContext.scala (hunk 5 above)
Comment (tdikland) [reply src/main/scala/com/databricks/labs/mosaic/functions/MosaicContext.scala:514]: Of course!

**Review by edurdevic — APPROVED** [review]
LGTM
